### ISSUE TITLE
Error message "400 / Bad Request / Internal Engine Error" when using backup.py on some VMs

### ISSUE THREAD

**Issue by fthommen** [wefixit-AT/oVirtBackup#16]
When using backup.py on some(!) VMs, we get a "Bad Request / Internal Engine Error" during the cloning step:

```
$ ./backup.py -c ./config_testvm.cfg -d
Apr 12 10:50:05: Start backup for: testvm
Apr 12 10:50:24: Snapshot creation started ...
Apr 12 10:50:32: Snapshot operation(creation) in progress ...
Apr 12 10:50:41: Snapshot operation(creation) in progress ...
Apr 12 10:50:51: Snapshot operation(creation) in progress ...
Apr 12 10:51:00: Snapshot created
Apr 12 10:51:04: Clone into VM started ...
Apr 12 10:51:07: !!! Got a RequestError:
status: 400
reason: Bad Request
detail: Internal Engine Error
Apr 12 10:51:07: All backups done
Apr 12 10:51:07: Backup failured for:
Apr 12 10:51:07:   testvm
Apr 12 10:51:07: Some errors occured during the backup, please check the 
log file
$
```
The used config file looks as follows:
```
$ cat config_testvm.cfg
[config]
vm_names: ["testvm"]
vm_middle=_BACKUP
snapshot_description=Snapshot for backup script
server=https://our.ovirt.server/ovirt-engine/api
username=xxxx@xxxx
password=xxxx
export_domain=our_export_domain
timeout=5
cluster_name=Default
backup_keep_count=3
dry_run=False
vm_name_max_length=32
storage_domain=our_storage_domain
storage_space_threshold=0.1
$
```
The same script with only the VM name changed in the configuration file works fine on other VMs.  When running the process for the failing VM manually through oVirt's admin webgui, then no error appears.

There is no obvious configuration difference between working and failing VMs.  The backup script is run directly on the oVirt server (`our.ovirt.server`).

System and oVirt version information:
```
$ cat /etc/os-release 
NAME="CentOS Linux"
VERSION="7 (Core)"
ID="centos"
ID_LIKE="rhel fedora"
VERSION_ID="7"
PRETTY_NAME="CentOS Linux 7 (Core)"
ANSI_COLOR="0;31"
CPE_NAME="cpe:/o:centos:centos:7"
HOME_URL="https://www.centos.org/"
BUG_REPORT_URL="https://bugs.centos.org/"

CENTOS_MANTISBT_PROJECT="CentOS-7"
CENTOS_MANTISBT_PROJECT_VERSION="7"
REDHAT_SUPPORT_PRODUCT="centos"
REDHAT_SUPPORT_PRODUCT_VERSION="7"

$ rpm -qa | grep ovirt
ovirt-engine-restapi-3.5.6.2-1.el7.centos.noarch
ovirt-hosted-engine-ha-1.2.8-1.el7.centos.noarch
ovirt-release35-006-1.noarch
ovirt-engine-setup-plugin-websocket-proxy-3.5.6.2-1.el7.centos.noarch
ovirt-engine-extensions-api-impl-3.5.6.2-1.el7.centos.noarch
ovirt-engine-sdk-python-3.6.3.0-1.el7.noarch
ovirt-image-uploader-3.5.1-1.el7.centos.noarch
ovirt-engine-backend-3.5.6.2-1.el7.centos.noarch
ovirt-engine-userportal-3.5.6.2-1.el7.centos.noarch
ovirt-engine-setup-base-3.5.6.2-1.el7.centos.noarch
ovirt-engine-jboss-as-7.1.1-1.el7.x86_64
ovirt-engine-setup-3.5.6.2-1.el7.centos.noarch
ovirt-engine-dbscripts-3.5.6.2-1.el7.centos.noarch
ovirt-engine-cli-3.5.0.6-1.el7.centos.noarch
ovirt-engine-3.5.6.2-1.el7.centos.noarch
ovirt-engine-setup-plugin-ovirt-engine-common-3.5.6.2-1.el7.centos.noarch
ovirt-host-deploy-1.3.2-1.el7.centos.noarch
ovirt-engine-setup-plugin-ovirt-engine-3.5.6.2-1.el7.centos.noarch
ovirt-engine-tools-3.5.6.2-1.el7.centos.noarch
ovirt-iso-uploader-3.5.2-1.el7.centos.noarch
ovirt-host-deploy-java-1.3.2-1.el7.centos.noarch
ovirt-engine-webadmin-portal-3.5.6.2-1.el7.centos.noarch
libgovirt-0.3.3-1.el7.x86_64
ovirt-hosted-engine-setup-1.2.6.1-1.el7.centos.noarch
ovirt-engine-lib-3.5.6.2-1.el7.centos.noarch
ovirt-engine-websocket-proxy-3.5.6.2-1.el7.centos.noarch
$
```

Is there a way to further debug this kind of problem?  Like a `-vvv` option?  Looking at the ovirt-engine log on the server doesn't really help, as the relevant messages are mixed with hundreds of other messages being printed during normal operation of the oVirt engine.  In the concrete case it would e.g. be helpful to track down the concrete request (with data), which is failing.  We are currently using commit `394a8743fc59a9bb7b4a131b133aa992fb3a1f30` of oVirtBackup

```
$ git rev-parse HEAD
394a8743fc59a9bb7b4a131b133aa992fb3a1f30
$
```
Thanks in advance
Frank Thommen

**Comment by fthommen** [comment]
Sorry, I forgot to document system type and oVirt versions:
```
$ cat /etc/os-release 
NAME="CentOS Linux"
VERSION="7 (Core)"
ID="centos"
ID_LIKE="rhel fedora"
VERSION_ID="7"
PRETTY_NAME="CentOS Linux 7 (Core)"
ANSI_COLOR="0;31"
CPE_NAME="cpe:/o:centos:centos:7"
HOME_URL="https://www.centos.org/"
BUG_REPORT_URL="https://bugs.centos.org/"

CENTOS_MANTISBT_PROJECT="CentOS-7"
CENTOS_MANTISBT_PROJECT_VERSION="7"
REDHAT_SUPPORT_PRODUCT="centos"
REDHAT_SUPPORT_PRODUCT_VERSION="7"

$ rpm -qa | grep ovirt
ovirt-engine-restapi-3.5.6.2-1.el7.centos.noarch
ovirt-hosted-engine-ha-1.2.8-1.el7.centos.noarch
ovirt-release35-006-1.noarch
ovirt-engine-setup-plugin-websocket-proxy-3.5.6.2-1.el7.centos.noarch
ovirt-engine-extensions-api-impl-3.5.6.2-1.el7.centos.noarch
ovirt-engine-sdk-python-3.6.3.0-1.el7.noarch
ovirt-image-uploader-3.5.1-1.el7.centos.noarch
ovirt-engine-backend-3.5.6.2-1.el7.centos.noarch
ovirt-engine-userportal-3.5.6.2-1.el7.centos.noarch
ovirt-engine-setup-base-3.5.6.2-1.el7.centos.noarch
ovirt-engine-jboss-as-7.1.1-1.el7.x86_64
ovirt-engine-setup-3.5.6.2-1.el7.centos.noarch
ovirt-engine-dbscripts-3.5.6.2-1.el7.centos.noarch
ovirt-engine-cli-3.5.0.6-1.el7.centos.noarch
ovirt-engine-3.5.6.2-1.el7.centos.noarch
ovirt-engine-setup-plugin-ovirt-engine-common-3.5.6.2-1.el7.centos.noarch
ovirt-host-deploy-1.3.2-1.el7.centos.noarch
ovirt-engine-setup-plugin-ovirt-engine-3.5.6.2-1.el7.centos.noarch
ovirt-engine-tools-3.5.6.2-1.el7.centos.noarch
ovirt-iso-uploader-3.5.2-1.el7.centos.noarch
ovirt-host-deploy-java-1.3.2-1.el7.centos.noarch
ovirt-engine-webadmin-portal-3.5.6.2-1.el7.centos.noarch
libgovirt-0.3.3-1.el7.x86_64
ovirt-hosted-engine-setup-1.2.6.1-1.el7.centos.noarch
ovirt-engine-lib-3.5.6.2-1.el7.centos.noarch
ovirt-engine-websocket-proxy-3.5.6.2-1.el7.centos.noarch
$
```
ft.

**Comment by wefixit-AT** [comment]
More debugging is not easy with the SDK/API, the exception which is given from the SDK is currently displayed, much more is not possible I think.

But I made a branch with some changes and added a traceback. Can you checkout the branch "issue_16" and post the output here. Check first if it doesn't contain sensitive information's !!! ;-)

If this helps, I will add a more debug output as you suggested. Sound's like a good idea.

**Comment by fthommen** [comment]
With this version I get
```
$ ./backup.py -c ./config_testvm.cfg -d
Apr 14 16:04:12: Start backup for: testvm
Apr 14 16:04:30: Found snapshots(1):
Apr 14 16:04:30: Snapshots description: Snapshot for backup script, Created on: 2016-04-12 13:15:34.983000+02:00
Apr 14 16:04:32: Snapshot deletion started ...
Apr 14 16:04:36: Snapshot operation(deletion) in progress ...
Apr 14 16:04:45: Snapshot operation(deletion) in progress ...
Apr 14 16:04:55: Snapshot operation(deletion) in progress ...
Apr 14 16:05:04: Snapshot operation(deletion) in progress ...
Apr 14 16:05:13: Snapshot operation(deletion) in progress ...
Apr 14 16:05:22: Snapshot operation(deletion) in progress ...
Apr 14 16:05:31: Snapshot operation(deletion) in progress ...
Apr 14 16:05:39: Snapshot operation(deletion) in progress ...
Apr 14 16:05:48: Snapshot operation(deletion) in progress ...
Apr 14 16:05:57: Snapshot operation(deletion) in progress ...
Apr 14 16:06:06: Snapshot operation(deletion) in progress ...
Apr 14 16:06:15: Snapshot operation(deletion) in progress ...
Apr 14 16:06:24: Snapshot operation(deletion) in progress ...
Apr 14 16:06:32: Snapshot operation(deletion) in progress ...
Apr 14 16:06:40: Snapshots deleted
Apr 14 16:06:41: Snapshot creation started ...
Apr 14 16:06:49: Snapshot operation(creation) in progress ...
Apr 14 16:06:58: Snapshot operation(creation) in progress ...
Apr 14 16:07:06: Snapshot created
Apr 14 16:07:20: Clone into VM started ...
Apr 14 16:07:23: !!! Got a RequestError: 
status: 400
reason: Bad Request
detail: Internal Engine Error
Traceback (most recent call last):
  File "./backup.py", line 127, in main
    api.vms.add(params.VM(name=vm_clone_name, memory=vm.get_memory(), cluster=api.clusters.get(config.get_cluster_name()), snapshots=snapshots_param))
  File "/usr/lib/python2.7/site-packages/ovirtsdk/infrastructure/brokers.py", line 35196, in add
    headers={"Correlation-Id":correlation_id, "Expect":expect}
  File "/usr/lib/python2.7/site-packages/ovirtsdk/infrastructure/proxy.py", line 75, in add
    return self.request('POST', url, body, headers)
  File "/usr/lib/python2.7/site-packages/ovirtsdk/infrastructure/proxy.py", line 115, in request
    persistent_auth=self.__persistent_auth
  File "/usr/lib/python2.7/site-packages/ovirtsdk/infrastructure/connectionspool.py", line 79, in do_request
    persistent_auth)
  File "/usr/lib/python2.7/site-packages/ovirtsdk/infrastructure/connectionspool.py", line 156, in __do_request
    raise errors.RequestError(response_code, response_reason, response_body)
RequestError: 
status: 400
reason: Bad Request
detail: Internal Engine Error
Apr 14 16:07:23: All backups done
Apr 14 16:07:23: Backup failured for:
Apr 14 16:07:23:   testvm
Apr 14 16:07:23: Some errors occured during the backup, please check the log file
$
```

The configuration is the same as before with an added `use_short_suffix = false`.  I have problems interpreting this traceback

**Comment by wefixit-AT** [comment]
Hi, its not really clear right now. So I wrote some check's to see if the config is correct. Can you checkout the latest changes on the branch "issue_16" and see if it prints some errors about missing config settings for export_domain, cluster_name and storage_domain. (I will refactor this later)

commit: 3e244711183976cec4faef4857f9474aaa00fcda

**Comment by fthommen** [comment]
I can't see a big difference:

```
$ ./backup.py -c ./config_testvm.cfg -d
Apr 15 11:09:06: Start backup for: testvm
Apr 15 11:09:24: Found snapshots(1):
Apr 15 11:09:24: Snapshots description: Snapshot for backup script, Created on: 2016-04-14 16:06:43.088000+02:00
Apr 15 11:09:26: Snapshot deletion started ...
Apr 15 11:09:30: Snapshot operation(deletion) in progress ...
Apr 15 11:09:39: Snapshot operation(deletion) in progress ...
Apr 15 11:09:48: Snapshot operation(deletion) in progress ...
Apr 15 11:09:58: Snapshot operation(deletion) in progress ...
Apr 15 11:10:07: Snapshot operation(deletion) in progress ...
Apr 15 11:10:16: Snapshot operation(deletion) in progress ...
Apr 15 11:10:26: Snapshot operation(deletion) in progress ...
Apr 15 11:10:35: Snapshot operation(deletion) in progress ...
Apr 15 11:10:44: Snapshot operation(deletion) in progress ...
Apr 15 11:10:52: Snapshot operation(deletion) in progress ...
Apr 15 11:11:01: Snapshot operation(deletion) in progress ...
Apr 15 11:11:09: Snapshot operation(deletion) in progress ...
Apr 15 11:11:18: Snapshot operation(deletion) in progress ...
Apr 15 11:11:27: Snapshot operation(deletion) in progress ...
Apr 15 11:11:35: Snapshots deleted
Apr 15 11:11:36: Snapshot creation started ...
Apr 15 11:11:43: Snapshot operation(creation) in progress ...
Apr 15 11:11:52: Snapshot operation(creation) in progress ...
Apr 15 11:12:01: Snapshot created
Apr 15 11:12:15: Clone into VM started ...
Apr 15 11:12:17: !!! Got a RequestError: 
status: 400
reason: Bad Request
detail: Internal Engine Error
Traceback (most recent call last):
  File "./backup.py", line 145, in main
    api.vms.add(params.VM(name=vm_clone_name, memory=vm.get_memory(), cluster=api.clusters.get(config.get_cluster_name()), snapshots=snapshots_param))
  File "/usr/lib/python2.7/site-packages/ovirtsdk/infrastructure/brokers.py", line 35196, in add
    headers={"Correlation-Id":correlation_id, "Expect":expect}
  File "/usr/lib/python2.7/site-packages/ovirtsdk/infrastructure/proxy.py", line 75, in add
    return self.request('POST', url, body, headers)
  File "/usr/lib/python2.7/site-packages/ovirtsdk/infrastructure/proxy.py", line 115, in request
    persistent_auth=self.__persistent_auth
  File "/usr/lib/python2.7/site-packages/ovirtsdk/infrastructure/connectionspool.py", line 79, in do_request
    persistent_auth)
  File "/usr/lib/python2.7/site-packages/ovirtsdk/infrastructure/connectionspool.py", line 156, in __do_request
    raise errors.RequestError(response_code, response_reason, response_body)
RequestError: 
status: 400
reason: Bad Request
detail: Internal Engine Error
Apr 15 11:12:17: All backups done
Apr 15 11:12:17: Backup failured for:
Apr 15 11:12:17:   testvm
Apr 15 11:12:17: Some errors occured during the backup, please check the log file
$
```
Can you say from this traceback, which step/function fails?

**Comment by wefixit-AT** [comment]
I see a difference :-) It means that the the values are correct.

Can you please attach the engine.log which is generated during the backup.

**Comment by fthommen** [comment]
It is impossible for me to see, which entries in the 40k-line logfile came from the backup job :-).  I'll rerun backup.py to generate new entries and filter them out as they are being generated.  However I'd prefer **not** to post them publicly.  Is it ok if I send them to you personally?  Do you want me to also send you the engine.log entries for a VM where backup.py runs fine?

**Comment by wefixit-AT** [comment]
Yes you can send it to me, I will handle it with care and don't publish sensitive Information's and wipe it after we find the problem.

Can you also create a empty third test VM with a disk of 1GB without a Operating System and run the backup on it. Maybe the VM has a strange problem.

I had this before on a test network with a VM where I hit some official Bug's (again) which where already reported to oVirt. There was the problem that a snapshot was generate but the snapshot disk was, for unknown reasons, corrupt and the backup-script did the job but the export VM has no disks stored. During the debugging I saw that all the SDK methods report no errors but there where thousand entries in the engine.log which are not forwarded through the SDK.

In short, currently I'm sceptical since I switched from VMWare ESXi to oVirt, before I had never problems. But this is another story :-(

**Comment by fthommen** [comment]
Logs and output are on the way to your mailbox.  As for the minimal test-VM, I'll only be able to do this early next week.  I'll then send you it's backup output as soon as it is available.

**Comment by fthommen** [comment]
Seems to have been a problem with some VMs themselves which could also not exported and imported properly through the oVirt WebUI.  After rebuilding the VMs from scratch, the backup problems have gone.  Thanks a lot for all your help.
